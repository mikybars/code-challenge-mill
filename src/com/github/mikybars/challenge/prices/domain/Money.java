package com.github.mikybars.challenge.prices.domain;

import java.math.BigDecimal;
import java.util.Currency;

public record Money(BigDecimal amount, Currency currency) {

}
