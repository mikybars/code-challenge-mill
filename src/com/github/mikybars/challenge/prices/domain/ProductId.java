package com.github.mikybars.challenge.prices.domain;

public record ProductId(String id) {

}
