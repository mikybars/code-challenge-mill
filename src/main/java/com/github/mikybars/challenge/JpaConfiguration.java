package com.github.mikybars.challenge;

import org.springframework.context.annotation.Configuration;
import org.springframework.data.jpa.repository.config.EnableJpaRepositories;

@Configuration
@EnableJpaRepositories(considerNestedRepositories = true)
public class JpaConfiguration {

}
