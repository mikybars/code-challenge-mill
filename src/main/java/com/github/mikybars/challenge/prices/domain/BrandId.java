package com.github.mikybars.challenge.prices.domain;

public record BrandId(String id) {

}
