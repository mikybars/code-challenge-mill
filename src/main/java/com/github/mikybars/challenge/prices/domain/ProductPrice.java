package com.github.mikybars.challenge.prices.domain;

import java.time.LocalDateTime;

public record ProductPrice(ProductId productId, BrandId brandId, PriceListId priceListId,
                           LocalDateTime startDate, LocalDateTime endDate,
                           Money price) {

}
