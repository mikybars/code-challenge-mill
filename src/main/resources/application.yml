spring:
  application:
    name: code-challenge
  config:
    import:
      - error-handling.yml