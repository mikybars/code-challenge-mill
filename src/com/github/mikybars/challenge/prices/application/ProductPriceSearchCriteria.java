package com.github.mikybars.challenge.prices.application;

import com.github.mikybars.challenge.prices.domain.ProductId;
import java.time.LocalDateTime;

public record ProductPriceSearchCriteria(LocalDateTime applicationDate, ProductId productId) {

}