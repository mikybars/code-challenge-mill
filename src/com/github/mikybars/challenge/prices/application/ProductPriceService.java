package com.github.mikybars.challenge.prices.application;

import com.github.mikybars.challenge.common.NotFoundException;
import com.github.mikybars.challenge.prices.application.ports.in.GetProductPriceUseCase;
import com.github.mikybars.challenge.prices.application.ports.out.ProductPriceRepository;
import com.github.mikybars.challenge.prices.domain.ProductPrice;
import java.util.function.Supplier;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
class ProductPriceService implements GetProductPriceUseCase {

  private final ProductPriceRepository productPriceRepository;

  @Override
  public ProductPrice execute(ProductPriceSearchCriteria searchCriteria) {
    return productPriceRepository
        .findTheHighestRankedBy(searchCriteria)
        .orElseThrow(notFoundExceptionTriggeredBy(searchCriteria));
  }

  private static Supplier<NotFoundException> notFoundExceptionTriggeredBy(
      ProductPriceSearchCriteria criteria) {
    return () -> new NotFoundException("no product price found for %s".formatted(criteria));
  }
}
