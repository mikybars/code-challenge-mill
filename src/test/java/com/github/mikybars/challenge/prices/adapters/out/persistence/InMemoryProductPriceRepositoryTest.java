package com.github.mikybars.challenge.prices.adapters.out.persistence;

import static org.assertj.core.api.Assertions.assertThat;

import com.github.mikybars.challenge.JpaConfiguration;
import com.github.mikybars.challenge.prices.application.ProductPriceSearchCriteria;
import com.github.mikybars.challenge.prices.domain.BrandId;
import com.github.mikybars.challenge.prices.domain.ProductId;
import com.github.mikybars.challenge.prices.domain.ProductPrice;
import java.time.LocalDateTime;
import java.util.Optional;
import org.approvaltests.JsonApprovals;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.context.annotation.Import;
import org.springframework.test.context.ActiveProfiles;

@DataJpaTest
@ActiveProfiles("standalone")
@Import({
    JpaConfiguration.class,
    InMemoryProductPriceRepository.class,
    ProductPricePersistenceMapperImpl.class
})
class InMemoryProductPriceRepositoryTest {

  static final LocalDateTime onlyPriceStartDate = LocalDateTime.parse("2020-06-14T00:00:00");
  static final LocalDateTime onlyPriceEndDate = LocalDateTime.parse("2020-12-31T23:59:59");
  static final LocalDateTime highestRankedStartDate = LocalDateTime.parse("2020-06-14T15:00:00");
  static final ProductId productWithOnePrice = new ProductId("35455");
  static final BrandId brandWithOnePrice = new BrandId("1");

  @Autowired
  InMemoryProductPriceRepository productPriceRepository;

  @Test
  void returnEmptyWhenDateIsBefore() {
    Optional<ProductPrice> productPrice = productPriceRepository.findTheHighestRankedBy(
        new ProductPriceSearchCriteria(onlyPriceStartDate.minusSeconds(1), productWithOnePrice
        ));

    assertThat(productPrice).isEmpty();
  }

  @Test
  void returnOneWhenDateEqualsStart() {
    Optional<ProductPrice> productPrice = productPriceRepository.findTheHighestRankedBy(
        new ProductPriceSearchCriteria(onlyPriceStartDate, productWithOnePrice));

    assertThat(productPrice).get()
        .satisfies(JsonApprovals::verifyAsJson);
  }

  @Test
  void returnOneWhenDateIsInBetween() {
    Optional<ProductPrice> productPrice = productPriceRepository.findTheHighestRankedBy(
        new ProductPriceSearchCriteria(onlyPriceStartDate.plusSeconds(1), productWithOnePrice));

    assertThat(productPrice).get()
        .extracting(ProductPrice::productId, ProductPrice::brandId)
        .containsExactly(productWithOnePrice, brandWithOnePrice);
  }

  @Test
  void returnOneWhenDateEqualsEnd() {
    Optional<ProductPrice> productPrice = productPriceRepository.findTheHighestRankedBy(
        new ProductPriceSearchCriteria(onlyPriceEndDate, productWithOnePrice));

    assertThat(productPrice).get()
        .extracting(ProductPrice::productId, ProductPrice::brandId)
        .containsExactly(productWithOnePrice, brandWithOnePrice);
  }

  @Test
  void returnEmptyWhenDateIsAfter() {
    Optional<ProductPrice> productPrice = productPriceRepository.findTheHighestRankedBy(
        new ProductPriceSearchCriteria(onlyPriceEndDate.plusSeconds(1), productWithOnePrice));

    assertThat(productPrice).isEmpty();
  }

  @Test
  void returnHighestRanked() {
    BrandId brandWithMultiplePrices = new BrandId("1");
    Optional<ProductPrice> productPrice = productPriceRepository.findTheHighestRankedBy(
        new ProductPriceSearchCriteria(highestRankedStartDate, productWithOnePrice));

    assertThat(productPrice).get()
        .extracting(ProductPrice::productId, ProductPrice::brandId, ProductPrice::startDate)
        .containsExactly(productWithOnePrice, brandWithMultiplePrices, highestRankedStartDate);
  }
}