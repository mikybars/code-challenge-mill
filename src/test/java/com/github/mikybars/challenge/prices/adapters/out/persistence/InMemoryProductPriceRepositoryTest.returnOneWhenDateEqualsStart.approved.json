{
  "productId": {
    "id": "35455"
  },
  "brandId": {
    "id": "1"
  },
  "priceListId": {
    "id": "1"
  },
  "startDate": "2020-06-14T00:00",
  "endDate": "2020-12-31T23:59:59",
  "price": {
    "amount": 35.50,
    "currency": "EUR"
  }
}